src/transformers/tokenization_utils.py

```diff
@@ -956,3 +956,19 @@ def _decode(
             return clean_text
         else:
             return text
+
+
+class SentencePieceStringConversionMixin:
+    def convert_tokens_to_string(self, tokens):
+        """Converts a sequence of tokens (string) in a single string."""
+        current_sub_tokens = []
+        out_string = ""
+        for token in tokens:
+            # make sure that special tokens are not decoded using sentencepiece model
+            if token in self.all_special_tokens:
+                out_string += self.sp_model.decode(current_sub_tokens) + token + " "
```

Comment:
This is related to my previous comment, but I'm not sure what the behavior should be here to add spaces. How did you choose this strategy?

Reply:
This is the exact code from PR #8435, I assumed this implementation would be relevant for all SPM tokenizers, but it may well be that this assumption is incorrect.

Reply:
After running the test you suggested, I do believe that this extra space is an error, at least compared to the Rust tokenizers. Removing the space passes the test 👍

Reply:
I won't be able to look at this issue `T5Tokenizer` with the again until next Monday but I promise to come back to you early next week

Reply:
Thank you very much! I will be waiting patiently 😄

Reply:
Actually this error with T5 makes me think that we probably have the same problem with the other tokenizers. Would you like to add a new test on `decode` that would check this?
 
I think that the intrication between `decode` and `convert_tokens_to_string` is very subtle and we need to be very careful about these behaviors. :smile:

Reply:
Will do!

Reply:
Thank you 🤩

Reply:
I have written the following test:
```python
def test_sentencepiece_tokenize_and_decode(self):
    if not self.test_sentencepiece:
        return

    text = "This is text to test the tokenizer."
    if self.test_rust_tokenizer:
        tokenizer = self.get_tokenizer()
        rust_tokenizer = self.get_rust_tokenizer()

        slow_ids = tokenizer(text).input_ids
        fast_ids = rust_tokenizer(text).input_ids
        self.assertEqual(slow_ids, fast_ids)

        slow_decoded = tokenizer.decode(slow_ids)
        fast_decoded = rust_tokenizer.decode(slow_ids)
        self.assertEqual(slow_decoded, fast_decoded)
```

The behaviors of slow and fast tokenizers in `decode` are indeed mismatched, and the mismatch is, to me, quite odd. Let's take a look at a few failures (these tokenizers pass `test_sentencepiece_tokenize_and_convert_tokens_to_string`):

|Tokenizer|Output|
|-|-|
|Albert|`'[CLS]this is text to test the tokenizer.[SEP]' != '[CLS] this is text to test the tokenizer.[SEP]'`|
|Barthez|`'<s>This is text to test the tokenizer.</s>' != '<s> This is text to test the tokenizer.</s>'`|
|BigBird|`'[CLS] This is text to test the tokenizer. [SEP]' != '[CLS] This is text to test the tokenizer.[SEP]'`|
|Camembert|`'<s>This is text to test the tokenizer.</s>' != '<s> This is text to test the tokenizer.</s>'`|
|MBart50|`'en_XXThis is text to test the tokenizer.</s>' != 'en_XX This is text to test the tokenizer.</s>'`|

It would seem that the correct behavior, as dictated by the fast tokenizers, is that the BOS token should have a space after it, but the EOS token comes right after the last token without a space.

This seems like a small thing to fix, but the question is whether the problem is that space, or is it something bigger that I'm not aware of.

Reply:
Hi! Quick update: I added the test for `decode` (could really use your input on this one 🙂) and fixed the tokenizers from the above table + some others. Most required a quick fix (basically: don't add a space if the previous token was a special token); BigBird, FNet, and XLNet required an override of `_decode` in addition to the fix that worked for the rest.

Your feedback is welcome whole-heartedly,  and have an excellent weekend!

Reply:
Hi @SaulLu 🙂 

I hope everything is okay with you, and that I'm not bothering you by tagging. It has been a while, so I'm just checking this PR doesn't "fall between the chairs" (I'm not even sure people say this in English 😅).

Have a wonderful day and I'm eagerly awaiting your reply!

Reply:
No problem, have a great weekend and we'll pick this up next week! 🙂

Reply:
So I really spent a lot of time thinking about your PR. I'm already impressed with everything you've done! :clap: 

The only thing that bothers me at the moment is the fact that we're including a lot of complexity in this RP to solve a problem (the output between the slow and the fast version is not the same) that is not the problem we wanted to solve initially. This second problem is in fact quite complex - beyond the case of sentencepiece-based tokenizers - and we are discussing it "internally" to see what we want to do (in the short and long term). So I'm sorry again, but I think it would be really beneficial to wait a bit longer to get more feedback. :disappointed: 

Regarding the initial problem, I was wondering if you managed to identify why the sentencepiece decode method ignored some tokens. Is it only the BOS, EOS and PAD type tokens?

Reply:
> Regarding the initial problem, I was wondering if you managed to identify why the sentencepiece decode method ignored some tokens. Is it only the BOS, EOS and PAD type tokens?

I haven't, as I didn't think we want to delve in the C++ implementation of SentencePiece. From what I understood, the code we added was only circumventing a problem with SP without touching the external dependency.

I have no problem waiting for further feedback, this PR and myself aren't planning on going anywhere anytime soon.

Looking forward to continue working with you, and I'm thankful to have you as my reviewer 🙂

Reply:
Thank you very much for your understanding!

> I haven't, as I didn't think we want to delve in the C++ implementation of SentencePiece. From what I understood, the code we added was only circumventing a problem with SP without touching the external dependency.

Yes, you're absolutely right, we are not going to touch the sentencepiece library at all, but it was more to make sure that there are no cases that we could miss. My guess is that sentencepiece model is just ignoring the bos_piece, eos_piece and pad_piece (listed [here](https://github.com/google/sentencepiece/blob/master/doc/options.md)) when decoding but I'm not 100% sure :smile:

Reply:
Hey @SaulLu 🙂 

Hope you're doing well and keeping healthy!

Since the issue is now marked as stale, I thought I'll check the status of this issue with you.

Have an excellent week!

Reply:
Hi @SaulLu,

How are you?

Any news regarding this issue?

Thanks 🙂

Reply:
I think we should split the decision in two:
1. Tokenizers that don't have a fast version, and then there's nothing to compare to, so users using `AutoTokenizer.from_pretrained()` even with `use_fast=True` will get a slow tokenizer
2. Tokenizers that have both versions. This part, I think, will require some investigation from HF to see whether there are use-cases of a slow tokenizer when a fast one is available. As a user, I do `pip install transformers` and get `tokenizers` as a dependency. Then `AutoTokenizer.from_pretrained()` by default gives a fast tokenizer, if one's available. Why would I want to use a slow tokenizer? Answering this question will motivate the decision on what to do with the slow tokenizers' code. It could be that the code can be ugly-but-working since it won't get touched too often, but if the fast tokenizers get changed frequently and we want to keep feature parity with the slow tokenizers, then maintainable code is probably a priority.

All the best, Lucile 🙂

Reply:
Hi @beneyal ,

As promised I've spent some time today diving back into your PR. In the end, I think that what you have proposed is a good solution given all the constraints that we have.

>  Why would I want to use a slow tokenizer?
To come back to one of the last questions you raised, we have slow versions for several reasons, including the fact that they are extremely close to the original code of the authors (this is the source of truth for building the fast tokenizer - however this source of truth disappears when some features have not been implemented in the original code, which is very often the case with added tokens), they are python versions so they are more readable by users and for backward compatibility reasons (e.g. there are options in the `decode` method of the slow tokenizers that are not available in the fasts). 

Before I do a new PR review, do you think you'll have time to resolve the merge conflicts? If you don't have the time, don't hesitate to tell me, I can take care of it and open a PR on your branch with the necessary changes. :smile: 

Thank you again for your patience and for having worked so hard on the code base!

Reply:
Hey @SaulLu,

I've fixed the merge conflicts and running all the tokenization tests right now (takes a while 😅), and I see that DeBERTa v2 is failing. Should I try and fix it?

Thank ***you*** for your tremendous help 🙂

Reply:
Thanks again for your responsiveness! Regarding DeBERTa v2, could you share with me the tests that fail?

Reply:
All of them 😅 Once the testing finishes (currently 50%), I'll know exactly why. For context, I accepted the changes in `main` over what was in the file of my branch. Didn't do anything else with that file.

![image](https://user-images.githubusercontent.com/3891274/165714381-556fa947-2f65-43dc-a86c-542c355e0cae.png)

Reply:
So, after feeling dumb for not noticing the `pytest -n auto` argument, which is why the tests took almost 3 hours to run, I found the problem: when accepting the change, it removed a constructor argument I passed. I fixed that, and now only one test is failing:

![image](https://user-images.githubusercontent.com/3891274/165733745-7fa5ce23-ff54-4aa8-a621-f900dbb54171.png)

This is weird, since using the "microsoft/deberta-v2-xlarge" tokenizer works on both slow and fast. I believe the problem is the test vocab. Is there something I can do to fix it, other than changing the test sentence (which seems quite drastic)?

Reply:
Interesting. This change breaks `test_get_vocab`, which makes sense, but `test_sentencepiece_tokenize_and_decode` at least now shows the "real" problem, which is what I fixed for the other slow tokenizers: 

```
 AssertionError: '[CLS]<unk>his is text to test the tokenizer.[SEP]' != '[CLS] <unk>his is text to test the tokenizer.[SEP]'
```

It's the spacing thing. For some tokenizers, it was a simple fix, for others, it required overriding `_decode`, but I can fix this. I'll need to check how the fast tokenizer works to mimic it space-between-special-tokens-wise.

Reply:
Hi @SaulLu,

I made DeBERTa v2 act like the fast tokenizer. It's very ad-hoc, since the behavior of the special tokens is a bit weird, but it passes all the tests 🙂 

Would love to hear your feedback!